lib/Doctrine/ORM/Tools/Pagination/Paginator.php

```diff
@@ -150,14 +151,16 @@ public function getIterator()
 
             $subQuery->setFirstResult($offset)->setMaxResults($length);
 
-            $ids = array_map('current', $subQuery->getScalarResult());
+            $foundIdRows = $subQuery->getScalarResult();
 
-            $whereInQuery = $this->cloneQuery($this->query);
             // don't do this for an empty id array
-            if (count($ids) === 0) {
+            if ($foundIdRows === []) {
                 return new \ArrayIterator([]);
             }
 
+            $whereInQuery = $this->cloneQuery($this->query);
+            $ids          = array_map('current', $foundIdRows);
+
             $this->appendTreeWalker($whereInQuery, WhereInWalker::class);
             $whereInQuery->setHint(WhereInWalker::HINT_PAGINATOR_ID_COUNT, count($ids));
```

Comment:
I know it's not really related to your fix, but shouldn't we keep the matched IDs as a hint too? Would increase the cohesion a bit more too.

Reply:
We might leave this for v3, though.

Reply:
Just for the record, #8797 works in the other direction to replace the `count()` with a bool, since the information is not actually needed (at this time) and it improves cache efficiency.